File: web/src/lib/platform.ts
import { invoke } from '@tauri-apps/api/core'

export type NativePlatform = 'android' | 'ios' | 'windows' | 'macos' | 'linux' | 'unknown'
export type RuntimeFamily = 'browser' | 'mobile' | 'desktop'
export type BackgroundMode = 'unsupported' | 'android-foreground-service' | 'ios-task-window'

export interface PlatformCapabilities {
  platform: NativePlatform
  runtime: RuntimeFamily
  windowControls: boolean
  appUpdates: boolean
  dragOut: boolean
  nativeFilePaths: boolean
  documentPicker: boolean
  secureKeyStore: boolean
  biometric: boolean
  backgroundMode: BackgroundMode
  maxConcurrentTransfers: number
}

const browserCapabilities: PlatformCapabilities = {
  platform: 'unknown',
  runtime: 'browser',
  windowControls: false,
  appUpdates: false,
  dragOut: false,
  nativeFilePaths: false,
  documentPicker: false,
  secureKeyStore: false,
  biometric: false,
  backgroundMode: 'unsupported',
  maxConcurrentTransfers: 2,
}

let capabilities = browserCapabilities

export function hasTauriRuntime(): boolean {
  return typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window
}

/** Load capabilities before React and persisted stores are initialized. */
export async function initializePlatform(): Promise<PlatformCapabilities> {
  if (!hasTauriRuntime()) {
    capabilities = browserCapabilities
    return capabilities
  }
  capabilities = await invoke<PlatformCapabilities>('platform_capabilities')
  return capabilities
}

export function getPlatformCapabilities(): PlatformCapabilities {
  return capabilities
}

export function isDesktopRuntime(): boolean {
  return capabilities.runtime === 'desktop'
}

export function isMobileRuntime(): boolean {
  return capabilities.runtime === 'mobile'
}

File: src-tauri/src/commands/platform.rs
//! Cross-platform runtime capability discovery.
//!
//! The frontend must not infer native capabilities from a user-agent string:
//! Android and iOS are Tauri runtimes too, but they do not expose the desktop
//! window, updater or drag-out commands.

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct PlatformCapabilities {
    pub platform: &'static str,
    pub runtime: &'static str,
    pub window_controls: bool,
    pub app_updates: bool,
    pub drag_out: bool,
    pub native_file_paths: bool,
    pub document_picker: bool,
    pub secure_key_store: bool,
    pub biometric: bool,
    pub background_mode: &'static str,
    pub max_concurrent_transfers: u8,
}

#[tauri::command]
pub(crate) fn platform_capabilities() -> PlatformCapabilities {
    PlatformCapabilities {
        platform: current_platform(),
        runtime: if cfg!(mobile) { "mobile" } else { "desktop" },
        window_controls: cfg!(desktop),
        app_updates: cfg!(desktop),
        drag_out: cfg!(desktop),
        native_file_paths: cfg!(desktop),
        document_picker: true,
        secure_key_store: cfg!(mobile),
        biometric: false,
        background_mode: if cfg!(target_os = "android") {
            "android-foreground-service"
        } else if cfg!(target_os = "ios") {
            "ios-task-window"
        } else {
            "unsupported"
        },
        max_concurrent_transfers: if cfg!(mobile) { 2 } else { 5 },
    }
}

pub(crate) const fn current_platform() -> &'static str {
    if cfg!(target_os = "android") {
        "android"
    } else if cfg!(target_os = "ios") {
        "ios"
    } else if cfg!(target_os = "windows") {
        "windows"
    } else if cfg!(target_os = "macos") {
        "macos"
    } else if cfg!(target_os = "linux") {
        "linux"
    } else {
        "unknown"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn host_capabilities_match_the_compiled_target() {
        let capabilities = platform_capabilities();
        assert_eq!(capabilities.runtime, "desktop");
        assert!(capabilities.window_controls);
        assert!(capabilities.app_updates);
        assert_eq!(capabilities.max_concurrent_transfers, 5);
        assert_ne!(capabilities.platform, "unknown");
    }
}
